src/node_file.cc

```diff
@@ -2727,6 +2730,270 @@ static void Mkdtemp(const FunctionCallbackInfo<Value>& args) {
+    if (pathname[i] != '%') continue;
+
+    if (first_percent == 0) {
+      first_percent = i;
+    }
+
+    if (pathname[i + 1] != '2' || (pathname[i + 2] | 0x20) != 102) continue;
+
+    env->isolate()->ThrowException(ERR_INVALID_FILE_URL_PATH(
+        env->isolate(), "File URL path must not include encoded / characters"));
+
+    return false;
+  }
+
+  *result_file_path = ada::unicode::percent_decode(pathname, pathname.size());
+
+  return true;
+}
+
+bool FileURLToPathImpl(Environment* env,
+                       ada::url_aggregator* file_url,
+                       std::string* result_file_path) {
+  if (file_url->type != ada::scheme::FILE) {
+    env->isolate()->ThrowException(ERR_INVALID_URL_SCHEME(env->isolate()));
+
+    return false;
+  }
+
+#if _WIN_32
+  return FileURLToPathWin32Impl(env,
+                                file_url->get_hostname(),
+                                file_url->get_pathname(),
+                                result_file_path);
+#else
+  return FileURLToPathPosixImpl(env,
+                                file_url->get_hostname(),
+                                file_url->get_pathname(),
+                                result_file_path);
+#endif
+}
+
+// this method skip the permission model, you MUST verify the permission
+// manually if you want to use this method
+bool UNSAFE_FileExist(Environment* env, const std::string& file_path) {
+  uv_fs_t req;
+
+  int rc = uv_fs_stat(env->event_loop(), &req, file_path.c_str(), nullptr);
+
+  if (rc == 0) {
+    const uv_stat_t* const s = static_cast<const uv_stat_t*>(req.ptr);
+    rc = !!(s->st_mode & S_IFDIR);
+  }
+
+  uv_fs_req_cleanup(&req);
+
+  // rc is 0 if the path refers to a file
+  return rc == 0;
+}
+
+void LegacyMainResolve(const FunctionCallbackInfo<Value>& args) {
+  CHECK_GE(args.Length(), 1);
+  // packageJSONUrl
+  // /home/user/project/package.json
+  CHECK(args[0]->IsString());
+
+  Environment* env = Environment::GetCurrent(args);
+
+  std::string_view base_path =
+      Utf8Value(env->isolate(), args[0].As<String>()).ToStringView();
+  auto base = ada::parse<ada::url_aggregator>(base_path);
+
+  if (!base) {
+    return args.GetReturnValue().Set(-2);
+  }
+
+  ada::url_aggregator* base_pointer = &base.value();
+  ada::url_aggregator* file_path_agg;
+  ada::result<ada::url_aggregator> file_path_url;
+  std::string initial_file_path;
+  std::string file_path;
+
+  // packageConfig.main
+  // ./index.js
+  // ./lib/index.js
+  if (args[1]->IsString()) {
+    std::string package_config_main =
+        Utf8Value(env->isolate(), args[1].As<String>()).ToString();
+
+    file_path_url = ada::parse<ada::url_aggregator>(
+        std::string("./") + package_config_main, base_pointer);
+
+    if (!file_path_url) {
+      return args.GetReturnValue().Set(-2);
+    }
+
+    file_path_agg = &file_path_url.value();
+
+    if (!FileURLToPathImpl(env, file_path_agg, &initial_file_path)) return;
```

Comment:
I really feel you are leaking memory here. Please, consider passing variables by reference instead of pointers or using a std container for that, such as `std::unique_ptr `

Reply:
Pass by reference 👍

Reply:
I think now is right.